parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,448 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+import os
+
+import torch
+from torch import nn
+
+from parlai.core.torch_agent import TorchAgent, Output
+from parlai.core.thread_utils import SharedTable
+from parlai.core.utils import round_sigfigs, padded_3d
+
+
+class TorchRankerAgent(TorchAgent):
+    @staticmethod
+    def add_cmdline_args(argparser):
+        TorchAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('TorchRankerAgent')
+        agent.add_argument(
+            '-cands', '--candidates', type=str, default='inline',
+            choices=['batch', 'inline', 'fixed', 'vocab'],
+            help='The source of candidates during training '
+                 '(see TorchRankerAgent._build_candidates() for details).')
+        agent.add_argument(
+            '-ecands', '--eval-candidates', type=str,
+            choices=['batch', 'inline', 'fixed', 'vocab'],
+            help='The source of candidates during evaluation (defaults to the same'
+                 'value as --candidates if no flag is given)')
+        agent.add_argument(
+            '-fcp', '--fixed-candidates-path', type=str,
+            help='A text file of fixed candidates to use for all examples, one '
+                 'candidate per line')
+        agent.add_argument(
+            '--fixed-candidate-vecs', type=str, default='reuse',
+            help="One of 'reuse', 'replace', or a path to a file with vectors "
+                 "corresponding to the candidates at --fixed-candidates-path. "
+                 "By default, a candidate vector file is generated with the same path "
```

Comment:
i think this should appear as an extension of the model,
not the candidates path, because several models can
use the same candidates...perhaps
<model_name>.<cands_name>

Reply:
For what it's worth, I've been using <model_file>.<cands_name> in other models